modules/calib3d/src/ippe.cpp

```diff
@@ -0,0 +1,1124 @@
+            y = objectPoints.at<Vec3d>(i)[1];
+            z = objectPoints.at<Vec3d>(i)[2];
+        }
+
+        if (abs(z) > IPPE_SMALL)
+        {
+            isOnZPlane = false;
+        }
+
+        xBar += x;
+        yBar += y;
+        zBar += z;
+
+        UZero.at<double>(0, i) = x;
+        UZero.at<double>(1, i) = y;
+        UZero.at<double>(2, i) = z;
+    }
+    xBar = xBar / static_cast<double>(n);
+    yBar = yBar / static_cast<double>(n);
+    zBar = zBar / static_cast<double>(n);
+
+    for (int i = 0; i < n; i++)
+    {
+        UZero.at<double>(0, i) -= xBar;
+        UZero.at<double>(1, i) -= yBar;
+        UZero.at<double>(2, i) -= zBar;
+    }
+
+    Matx44d MCenter = Matx44d::eye();
+    MCenter(0, 3) = -xBar;
+    MCenter(1, 3) = -yBar;
+    MCenter(2, 3) = -zBar;
+
+    if (isOnZPlane)
+    {
+        //MmodelPoints2Canonical is given by MCenter
+        Mat(MCenter, false).copyTo(_MmodelPoints2Canonical);
+        for (int i = 0; i < n; i++)
+        {
+            canonicalObjPoints.at<Vec2d>(i)[0] = UZero.at<double>(0, i);
+            canonicalObjPoints.at<Vec2d>(i)[1] = UZero.at<double>(1, i);
+        }
+    }
+    else
+    {
+        Mat UZeroAligned(3, n, CV_64FC1);
+        Mat R; //rotation that rotates objectPoints to the plane z=0
+
+        if (!computeObjextSpaceR3Pts(objectPoints,R))
+        {
+            //we could not compute R, problably because there is a duplicate point in {objectPoints(0),objectPoints(1),objectPoints(2)}.
+            //So we compute it with the SVD (which is slower):
+            computeObjextSpaceRSvD(UZero,R);
+        }
+
+        UZeroAligned = R * UZero;
+
+        for (int i = 0; i < n; i++)
+        {
+            canonicalObjPoints.at<Vec2d>(i)[0] = UZeroAligned.at<double>(0, i);
+            canonicalObjPoints.at<Vec2d>(i)[1] = UZeroAligned.at<double>(1, i);
+//            assert(abs(UZeroAligned.at<double>(2, i)) <= IPPE_SMALL);
+            if (abs(UZeroAligned.at<double>(2, i)) > IPPE_SMALL)
+                CV_Error(Error::StsNoConv, "Cannot transform object points to the plane z=0!");
+        }
+
+        Mat MRot = Mat::zeros(4, 4, CV_64FC1);
+        MRot.at<double>(3, 3) = 1;
+
+        R.copyTo(MRot.colRange(0, 3).rowRange(0, 3));
+        Mat Mb = MRot * MCenter;
+        Mb.copyTo(_MmodelPoints2Canonical);
+    }
+}
+
+void PoseSolver::evalReprojError(InputArray _objectPoints, InputArray _imagePoints, InputArray _M, float& err)
+{
+    Mat projectedPoints;
+    Mat imagePoints = _imagePoints.getMat();
+    Mat r;
+    rot2vec(_M.getMat().colRange(0, 3).rowRange(0, 3), r);
+
+    Mat K = Mat::eye(3, 3, CV_64FC1);
+    Mat dist;
+    projectPoints(_objectPoints, r, _M.getMat().colRange(3, 4).rowRange(0, 3), K, dist, projectedPoints);
+
+    err = 0;
+    int n = _objectPoints.rows() * _objectPoints.cols();
+
+    float dx, dy;
+    for (int i = 0; i < n; i++)
+    {
+        if (projectedPoints.depth() == CV_32F)
```

Comment:
> projectedPoints.depth() 

Probably it is better to cache this value into some variable outside of the loop.